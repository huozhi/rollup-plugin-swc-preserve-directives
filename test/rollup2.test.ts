import { runTests } from './index.test'
import { rollup } from 'rollup2';

describe('preserve-directive (rollup 2)', () => {
  // @ts-expect-error -- rollup 2 type is incompatible w/
  runTests(rollup, 2);
});

