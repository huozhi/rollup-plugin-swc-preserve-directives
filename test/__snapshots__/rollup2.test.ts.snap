// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preserve-directive (rollup 2) fix string constant introduced by rollup commonjs 1`] = `
{
  "index": "import require$$0 from 'react';

function getDefaultExportFromCjs (x) {
	return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, 'default') ? x['default'] : x;
}

const { useState } = require$$0;
var commonjsVirtualModules = function client() {
    return useState(null);
};
var index = /*@__PURE__*/ getDefaultExportFromCjs(commonjsVirtualModules);

export { index as default };
",
}
`;

exports[`preserve-directive (rollup 2) issue #9 1`] = `
{
  "index": "'use strict';
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */ 
function emptyFunction() {}
function emptyFunctionWithReset() {}
emptyFunctionWithReset.resetWarningCache = emptyFunction;
module.exports = function() {
    function shim(props, propName, componentName, location, propFullName, secret) {
        var err = new Error('Calling PropTypes validators directly is not supported by the \`prop-types\` package. ' + 'Use PropTypes.checkPropTypes() to call them. ' + 'Read more at http://fb.me/use-check-prop-types');
        err.name = 'Invariant Violation';
        throw err;
    }
    shim.isRequired = shim;
    function getShim() {
        return shim;
    }
    // Important!
    // Keep this list in sync with production version in \`./factoryWithTypeCheckers.js\`.
    var ReactPropTypes = {
        array: shim,
        bigint: shim,
        bool: shim,
        func: shim,
        number: shim,
        object: shim,
        string: shim,
        symbol: shim,
        any: shim,
        arrayOf: getShim,
        element: shim,
        elementType: shim,
        instanceOf: getShim,
        node: shim,
        objectOf: getShim,
        oneOf: getShim,
        oneOfType: getShim,
        shape: getShim,
        exact: getShim,
        checkPropTypes: emptyFunctionWithReset,
        resetWarningCache: emptyFunction
    };
    ReactPropTypes.PropTypes = ReactPropTypes;
    return ReactPropTypes;
};
",
}
`;

exports[`preserve-directive (rollup 2) should merge duplicated directives 1`] = `
{
  "index": "'use client';
'use sukka';
const foo = 'foo';

const bar = 'bar';

const baz = 'baz';

export { bar, baz, foo };
",
}
`;

exports[`preserve-directive (rollup 2) should not output shebang to other entry chunks 1`] = `
{
  "cli": "#!/usr/bin/env node
console.log('index');
",
  "client": "'use client';
const foo = 'foo';

function client() {
    return 'client' + foo;
}

export { client };
",
  "server": "'use server';
function server() {
    return 'server';
}

export { server };
",
}
`;

exports[`preserve-directive (rollup 2) should output separate directive for multiple output chunk 1`] = `
{
  "client": "'use client';
'use sukka';
import { useState } from 'react';

const foo = 'is-client';

const ClientComponent = ()=>{
    const [count] = useState(0);
    return /*#__PURE__*/ React.createElement("div", null, "count: ", count);
};

export { ClientComponent, foo };
",
  "server": "'use server';
import { readFile } from 'fs/promises';

const ServerComponent = async ()=>{
    const foo = await readFile('__virtual_data/foo.txt', 'utf8');
    return /*#__PURE__*/ React.createElement("div", null, foo);
};

export { ServerComponent };
",
}
`;

exports[`preserve-directive (rollup 2) should preserve directives 1`] = `
{
  "index": "'use client';
import { useState } from 'react';

function client() {
    return useState(null);
}

export { client as default };
",
}
`;

exports[`preserve-directive (rollup 2) should preserve shebang 1`] = `
{
  "index": "#!/usr/bin/env node
console.log('shebang');
",
}
`;

exports[`preserve-directive (rollup 4) fix string constant introduced by rollup commonjs 1`] = `
{
  "index": "import require$$0 from 'react';

function getDefaultExportFromCjs (x) {
	return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, 'default') ? x['default'] : x;
}

const { useState } = require$$0;
var commonjsVirtualModules = function client() {
    return useState(null);
};
var index = /*@__PURE__*/ getDefaultExportFromCjs(commonjsVirtualModules);

export { index as default };
",
}
`;

exports[`preserve-directive (rollup 4) issue #9 1`] = `
{
  "index": "'use strict';
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */ 
function emptyFunction() {}
function emptyFunctionWithReset() {}
emptyFunctionWithReset.resetWarningCache = emptyFunction;
module.exports = function() {
    function shim(props, propName, componentName, location, propFullName, secret) {
        var err = new Error('Calling PropTypes validators directly is not supported by the \`prop-types\` package. ' + 'Use PropTypes.checkPropTypes() to call them. ' + 'Read more at http://fb.me/use-check-prop-types');
        err.name = 'Invariant Violation';
        throw err;
    }
    shim.isRequired = shim;
    function getShim() {
        return shim;
    }
    // Important!
    // Keep this list in sync with production version in \`./factoryWithTypeCheckers.js\`.
    var ReactPropTypes = {
        array: shim,
        bigint: shim,
        bool: shim,
        func: shim,
        number: shim,
        object: shim,
        string: shim,
        symbol: shim,
        any: shim,
        arrayOf: getShim,
        element: shim,
        elementType: shim,
        instanceOf: getShim,
        node: shim,
        objectOf: getShim,
        oneOf: getShim,
        oneOfType: getShim,
        shape: getShim,
        exact: getShim,
        checkPropTypes: emptyFunctionWithReset,
        resetWarningCache: emptyFunction
    };
    ReactPropTypes.PropTypes = ReactPropTypes;
    return ReactPropTypes;
};
",
}
`;

exports[`preserve-directive (rollup 4) should merge duplicated directives 1`] = `
{
  "index": "'use client';
'use sukka';
const foo = 'foo';

const bar = 'bar';

const baz = 'baz';

export { bar, baz, foo };
",
}
`;

exports[`preserve-directive (rollup 4) should not output shebang to other entry chunks 1`] = `
{
  "cli": "#!/usr/bin/env node
console.log('index');
",
  "client": "'use client';
const foo = 'foo';

function client() {
    return 'client' + foo;
}

export { client };
",
  "server": "'use server';
function server() {
    return 'server';
}

export { server };
",
}
`;

exports[`preserve-directive (rollup 4) should output separate directive for multiple output chunk 1`] = `
{
  "client": "'use client';
'use sukka';
import { useState } from 'react';

const foo = 'is-client';

const ClientComponent = ()=>{
    const [count] = useState(0);
    return /*#__PURE__*/ React.createElement("div", null, "count: ", count);
};

export { ClientComponent, foo };
",
  "server": "'use server';
import { readFile } from 'fs/promises';

const ServerComponent = async ()=>{
    const foo = await readFile('__virtual_data/foo.txt', 'utf8');
    return /*#__PURE__*/ React.createElement("div", null, foo);
};

export { ServerComponent };
",
}
`;

exports[`preserve-directive (rollup 4) should preserve directives 1`] = `
{
  "index": "'use client';
import { useState } from 'react';

function client() {
    return useState(null);
}

export { client as default };
",
}
`;

exports[`preserve-directive (rollup 4) should preserve shebang 1`] = `
{
  "index": "#!/usr/bin/env node
console.log('shebang');
",
}
`;
